hibernate-core/src/main/java/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java

```diff
@@ -1408,6 +1431,44 @@ public void pruneForSubclasses(TableGroup tableGroup, Map<String, EntityNameUse>
 		else {
 			tableReferenceJoins.removeIf( join -> !retainedTableReferences.contains( join.getJoinedTableReference() ) );
 		}
+
+		if ( needsTreatDiscriminator ) {
+			// We have to apply the discriminator condition to the root table reference join
+			boolean applied = applyDiscriminatorPredicate(
```

Comment:
It seems to me that this must always succeed, as `needsTreatDiscriminator` can only be `true` if `needsDiscriminator()` returns `true` and in `AbstractEntityPersister#createRootTableGroup`, we always create the primary table reference with the `getRootTableName()` if `needsDiscriminator() == true`.

So IMO the whole block in this if should be simplified to

```java
			if ( tableReferenceJoins.isEmpty() ) {
				final String discriminatorPredicate = getPrunedDiscriminatorPredicate( entityNameUses, metamodel, "t" );
				if ( discriminatorPredicate != null ) {
					final NamedTableReference tableReference = (NamedTableReference) tableGroup.getPrimaryTableReference();
					tableReference.setPrunedTableExpression( "(select * from " + getRootTableName() + " t where " + discriminatorPredicate + ")" );
				}
			}
			else {
				final TableReferenceJoin join = tableReferenceJoins.get( 0 );
				assert join.getJoinType() == SqlAstJoinType.INNER : "Found table reference join with root table of non-INNER type: " + join.getJoinType();
				final String discriminatorPredicate = getPrunedDiscriminatorPredicate(
						entityNameUses,
						metamodel,
						tableGroup.getPrimaryTableReference().getIdentificationVariable()
				);
				join.applyPredicate( new SqlFragmentPredicate( discriminatorPredicate ) );
			}
```

Reply:
This is not always true, as the table gorup might be created by a `LazyTableGroup` supplier like `ToOneAttributeMapping#createTableGroupInternal` which does not use the root table name. You probably don't remember but in my first iteration I also supposed the first join to always be the one with the root table, but our test suite confirmed that's often not the case.

Reply:
Ok, but can `tableReferenceJoins` be empty or i.e. reasonably not contain the root table?

Reply:
Well, if it's empty the `getPrimaryTableReference()` will always return the root table (I already use that in my code). it will also always contain the root table, but to find the correct join we have to check all `tableReferenceJoins` until we find it like I do in the for loop.

Reply:
The code you pointed me to in `ToOneAttributeMapping#createTableGroupInternal` suggests that if we have a mapping like:

```java
@ManyToOne
Subclass association;
```

That the root table might not be part of the joins for a query e.g. `select 1 from EntityA a join a.association`.

Reply:
Well, if that's the case we wouldn't need the discriminator to be applied, since the subtype's table would be directly joined. So maybe I just need to remove the `assert applied`?